src/api/source/mvt.js

```diff
@@ -68,70 +68,64 @@ export default class MVT extends Base {
         }
         const tile = new VectorTile(new Protobuf(arrayBuffer));
         const mvtLayer = tile.layers[Object.keys(tile.layers)[0]];
-
-        const { points, featureGeometries, properties } = this._decodeMVTLayer(mvtLayer, this._metadata, MVT_EXTENT);
-
-
+        const { geometries, properties } = this._decodeMVTLayer(mvtLayer, this._metadata, MVT_EXTENT);
         const rs = rsys.getRsysFromTile(x, y, z);
-        const dataframeGeometry = this._metadata.geomType == geometryTypes.POINT ? points : featureGeometries;
-        const dataframe = this._generateDataFrame(rs, dataframeGeometry, properties, mvtLayer.length, this._metadata.geomType);
+        const dataframe = this._generateDataFrame(rs, geometries, properties, mvtLayer.length, this._metadata.geomType);
+
         return dataframe;
     }
 
 
     _decodeMVTLayer(mvtLayer, metadata, mvt_extent) {
-        let points;
-        if (metadata.geomType == geometryTypes.POINT) {
-            points = new Float32Array(mvtLayer.length * 2);
+        switch (metadata.geomType) {
+            case geometryTypes.POINT:
+                return this._decode(mvtLayer, metadata, mvt_extent);
+            case geometryTypes.LINE:
+                return this._decode(mvtLayer, metadata, mvt_extent, decodeLines);
+            case geometryTypes.POLYGON:
+                return this._decode(mvtLayer, metadata, mvt_extent, decodePolygons);
+            default:
+                throw new Error('MVT: invalid geometry type');
         }
-        let featureGeometries = [];
-        const decodedProperties = {};
-        const decodingPropertyNames = Object.keys(metadata.properties).
-            filter(propertyName => metadata.properties[propertyName].type != 'geometry').
-            map(propertyName => metadata.properties[propertyName].sourceName || propertyName);
-        decodingPropertyNames.forEach(propertyName => {
-            decodedProperties[propertyName] = new Float32Array(mvtLayer.length + RTT_WIDTH);
-        });
+    }
+
+    _decode(mvtLayer, metadata, mvt_extent, decodeFn) {
+        const { properties, propertyNames } = this._getPropertyNames(metadata, mvtLayer.length);
+        const geometries = decodeFn ? [] : new Float32Array(mvtLayer.length * 2);
+
         for (let i = 0; i < mvtLayer.length; i++) {
             const f = mvtLayer.feature(i);
             const geom = f.loadGeometry();
-            const mvtGeomType = f.type;
-            if (metadata.geomType === undefined) {
```

Comment:
The geometry type was being auto-discovered, I think you just removed that functionality.

Reply:
Because the geometry comes in the metadata and currently is the same for all features, at least for windshaft.

For the MVT-Source I dont know the details but I see two options: auto-discover or force the user to indicate the geometry type in the metadata. In any case with the current implementation the geometry type would be the one given by the last feature and I'm not sure if this is fine.

Reply:
Yes, it is the case for Windshaft, but it is not the case for general MVT, and yes, if mixed geometry types are present things will be wrong, that's why I said that that's the real problem

Reply:
> The geometry type was being auto-discovered, I think you just removed that functionality.

Fixed :)

Reply:
> If mixed geometry types are present things will be wrong

Are you ok if we throw an error (at least while not implement this right)

Reply:
Yes, of course